chromium_src/chrome/browser/ui/webui/tab_search/tab_search_page_handler.h

```diff
@@ -0,0 +1,101 @@
+#define MaybeShowUI                      \
+  NotUsed();                             \
+  friend TabSearchPageHandler_BraveImpl; \
+  void MaybeShowUI
+
+#include "src/chrome/browser/ui/webui/tab_search/tab_search_page_handler.h"  // IWYU pragma: export
+
+#undef TabSearchPageHandler
+#undef MaybeShowUI
+
+namespace ai_chat {
+struct Tab;
+namespace mojom {
+enum class APIError;
+}  // namespace mojom
+}  // namespace ai_chat
+
+// Overrides TabSearchPageHandler to provide Brave-specific functionality.
+// See tab_search.mojom in chromium_src for our extended interface. Currently
+// it provides APIs needed for our tab organization feature using Leo.
+class TabSearchPageHandler : public TabSearchPageHandler_ChromiumImpl {
+ public:
+  // Used to store info for undo focus tabs.
+  struct TabInfo {
+    int tab_id;
+    int index;
+  };
+
+  TabSearchPageHandler(
+      mojo::PendingReceiver<tab_search::mojom::PageHandler> receiver,
+      mojo::PendingRemote<tab_search::mojom::Page> page,
+      content::WebUI* web_ui,
+      TopChromeWebUIController* webui_controller,
+      MetricsReporter* metrics_reporter);
+  ~TabSearchPageHandler() override;
+  TabSearchPageHandler(const TabSearchPageHandler&) = delete;
+  TabSearchPageHandler& operator=(const TabSearchPageHandler&) = delete;
+
+  // tab_search::mojom::PageHandler:
+  void GetSuggestedTopics(GetSuggestedTopicsCallback callback) override;
+  void GetFocusTabs(const std::string& topic,
+                    GetFocusTabsCallback callback) override;
+  void UndoFocusTabs(UndoFocusTabsCallback callback) override;
+  void OpenLeoGoPremiumPage() override;
+
+  void SetAIChatEngineForTesting(
+      std::unique_ptr<ai_chat::EngineConsumer> ai_chat_engine) {
+    ai_chat_engine_ = std::move(ai_chat_engine);
+  }
+
+  ai_chat::EngineConsumer* GetAIChatEngineForTesting() {
+    return ai_chat_engine_.get();
+  }
+
+  void SetFocusTabsInfoForTesting(
+      const base::flat_map<SessionID, std::vector<TabInfo>>& focus_tabs_info) {
+    focus_tabs_info_ = focus_tabs_info;
+  }
+
+ private:
+  void OnGetFocusTabs(GetFocusTabsCallback callback,
+                      base::expected<std::vector<std::string>,
+                                     ai_chat::mojom::APIError> result);
+  void OnGetSuggestedTopics(GetSuggestedTopicsCallback callback,
+                            base::expected<std::vector<std::string>,
+                                           ai_chat::mojom::APIError> result);
+  void OnTabOrganizationFeaturePrefChanged(Profile* profile);
+  tab_search::mojom::ErrorPtr GetError(ai_chat::mojom::APIError error);
+
+  ai_chat::EngineConsumer* GetAIEngineForTabOrganization();
+  std::vector<ai_chat::Tab> GetTabsForAIEngine();
+
+  // Map from session id to the list of tab info for undo last get focus tabs
+  // action (i.e. move tabs from the new window back to their original
+  // positions).
+  base::flat_map<SessionID, std::vector<TabInfo>> focus_tabs_info_;
```

Comment:
So basically:
1. We put the SessionId for the window into this map, with all the tabs it used to have
2. Then, when we undo we restore from this?

`SessionId` is tied to a Window then?

Reply:
Tied to a window, yes, it's from `browser->session_id()`.
key: browser->session_id() (serves as window ID here) 
value: tab infos that we moved out from this window when doing GetFocusTab.
Then undo used this to move them all back, yes.